src/url.rs

```diff
@@ -0,0 +1,149 @@
+use idna::punycode::decode_to_string;
+use pyo3::prelude::*;
+
+#[pyclass(name = "Url", module = "pydantic_core._pydantic_core")]
+#[derive(Clone)]
+#[cfg_attr(debug_assertions, derive(Debug))]
+pub struct PyUrl {
+    lib_url: url::Url,
+}
+
+static PUNYCODE_PREFIX: &str = "xn--";
+
+impl PyUrl {
+    pub fn new(lib_url: url::Url) -> Self {
+        Self { lib_url }
+    }
+
+    pub fn into_url(self) -> url::Url {
+        self.lib_url
+    }
+
+    fn decode_punycode(&self, domain: &str) -> Option<String> {
+        let mut result = String::with_capacity(domain.len());
+        for chunk in domain.split('.') {
+            if let Some(stripped) = chunk.strip_prefix(PUNYCODE_PREFIX) {
+                result.push_str(&decode_to_string(stripped)?);
+            } else {
+                result.push_str(chunk);
+            }
+            result.push('.');
```

Comment:
aren't there some cases where it might reallocate the string because the capacity is not enough?

Reply:
Possibly, but unlikely since the punycode representation should always be longer than the "unicode" (is there a better name?) representation.

Even if it does, it shouldn't cause an error?

Reply:
(@PrettyWood just mentioning it here since you not have seen, I just sent you a message on twitter, answer on https://github.com/pydantic/pydantic/issues/4665 would be great when you have a chance, sorry to hassle :bow:)